cpp/src/aztec/honk/commitment_scheme/ipa/ipa.hpp

```diff
@@ -0,0 +1,266 @@
+                              const std::vector<affine_element>& G_vector)
+    {
+        IpaProof proof;
+        auto& challenge_point = ipa_pub_input.challenge_point;
+        ASSERT(challenge_point != 0 && "The challenge point should not be zero");
+        const size_t poly_degree = ipa_pub_input.poly_degree;
+        // To check poly_degree is greater than zero and a power of two
+        ASSERT((poly_degree > 0) && (!(poly_degree & (poly_degree - 1))) &&
+               "The poly_degree should be positive and a power of two");
+        auto& aux_generator = ipa_pub_input.aux_generator;
+        auto a_vec = polynomial;
+        std::vector<affine_element> G_vec_local(poly_degree);
+        for (size_t i = 0; i < poly_degree; i++) {
+            G_vec_local[i] = G_vector[i];
+        }
+        // Construct b vector
+        std::vector<Fr> b_vec(poly_degree);
+        Fr b_power = 1;
+        for (size_t i = 0; i < poly_degree; i++) {
+            b_vec[i] = b_power;
+            b_power *= challenge_point;
+        }
+        // Iterate for log_2(poly_degree) rounds to compute the round commitments.
+        const size_t log_poly_degree = static_cast<size_t>(numeric::get_msb(poly_degree));
+        std::vector<element> L_elements(log_poly_degree);
+        std::vector<element> R_elements(log_poly_degree);
+        size_t round_size = poly_degree;
+
+        for (size_t i = 0; i < log_poly_degree; i++) {
+            round_size >>= 1;
+            // Compute inner_prod_L := < a_vec_lo, b_vec_hi > and inner_prod_R := < a_vec_hi, b_vec_lo >
+            Fr inner_prod_L = Fr::zero();
+            Fr inner_prod_R = Fr::zero();
+            for (size_t j = 0; j < round_size; j++) {
+                inner_prod_L += a_vec[j] * b_vec[round_size + j];
+                inner_prod_R += a_vec[round_size + j] * b_vec[j];
+            }
+            // L_i = < a_vec_lo, G_vec_hi > + inner_prod_L * aux_generator
+            auto pippenger_runtime_state = barretenberg::scalar_multiplication::pippenger_runtime_state(round_size);
+            element partial_L = barretenberg::scalar_multiplication::pippenger_without_endomorphism_basis_points(
+                &a_vec[0], &G_vec_local[round_size], round_size, pippenger_runtime_state);
+            partial_L += aux_generator * inner_prod_L;
+
+            // R_i = < a_vec_hi, G_vec_lo > + inner_prod_R * aux_generator
+            element partial_R = barretenberg::scalar_multiplication::pippenger_without_endomorphism_basis_points(
+                &a_vec[round_size], &G_vec_local[0], round_size, pippenger_runtime_state);
+            partial_R += aux_generator * inner_prod_R;
+
+            std::vector<element> round_commitments(2);
+            round_commitments[0] = partial_L;
+            round_commitments[1] = partial_R;
+            element::batch_normalize(&round_commitments[0], 2);
+
+            L_elements[i] = affine_element({ round_commitments[0].x, round_commitments[0].y });
```

Comment:
you can directly construct `affine_element` objects from `element`.


e.g. `L_elements[I] = affine_element(round_commitments[0])`

Reply:
The code would be cleaner (with a negligible performance effect) if we wrote:
```
L_elements[i] = affine_element(partial_L);
R_elements[i] = affine_element(partial_R);
```

Reply:
Thanks for this. Works.